Fix REST api

OGAWAHirofumi opened the pull request:
This PR fixes the issues in REST api path and adds tests to prevent regression.

And REST api error response is consolidated to rest_errorhandler.error_message()
for consistency, and easier to test.

(there are 2 PRs that is depending on this)

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+def _upload(client, data, token=None, filename=None, ftype=None, lifetime=None,
+            range_str=None, trans_id=None, no_encode=False):
+    if data:
+        if no_encode:
+            payload = data
+        else:
+            payload = base64.b64encode(data)
```
payload = base64.b64encode(data) if encode else data

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+def check_detail_or_download(app, client, fid, meta, data, download=False):
+    url = RestUrl(fid=fid)
+
+    if download:
+        url = url.download
+    else:
+        url = url.detail
```
url = url.download if download else url.detail

ThomasWaldmann replied:
(also fix other places)

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+    check_detail_or_download(app, client, fid, meta, UPLOAD_DATA)
+    check_detail_or_download(app, client, fid, meta, UPLOAD_DATA, download=True)
+
+    # upload again with not append position
```
?

OGAWAHirofumi replied:
changed comment

```python
# upload again with item.data.size != start position (invalid position)
```

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+
+class TmpConfig:
+    def __init__(self, app, tmp_config={}):
+        self.config = app.config
+        self.orig_config = {}
+        self.tmp_config = tmp_config
+        for k, v in tmp_config.items():
+            self.orig_config[k] = self.config[k]
+
+    def __enter__(self):
+        for k, v in self.tmp_config.items():
+            self.config[k] = v
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        for k, v in self.orig_config.items():
+            self.config[k] = self.orig_config[k]
```
using dict() to copy assumes that you do not have a need for `copy.deepcopy`.

OGAWAHirofumi replied:
Hm, we want to backup only part of app.config that specified keys by tmp_config.keys(), so not sure how use dict() or copy.deepcopy.

fixed to use dict.update()

ThomasWaldmann replied:
not sure if doing a partial backup in a python loop is faster than just creating a full backup by copying within the python interpreter. the latter would be easier in any case.

ThomasWaldmann replied:
(also: guess we rather optimize for simplicity here, not for performance)

OGAWAHirofumi replied:
Hm, I did want to change app.config's dict itself. because flask may have another reference to app.config's dict internally (or future).

e.g.
```python
some_context.config = app.config
...
if some_context.config['FOO']:
    ....
```

so I was not sure if the setting new dict to app.config works or not.

OGAWAHirofumi replied:
ah, i noticed no issue now. so changed to

```python
class TmpConfig:
    def __init__(self, app, tmp_config=None):
        self.config = app.config
        self.orig_config = copy.deepcopy(app.config)
        self.tmp_config = {} if tmp_config is None else tmp_config

    def __enter__(self):
        self.config.update(self.tmp_config)

    def __exit__(self, exc_type, exc_value, traceback):
        self.config.update(self.orig_config)
```

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+            self.config[k] = self.orig_config[k]
+
+
+class RestUrl:
```
is this really needed or could you also use flask's `url_for` to build URLs?

OGAWAHirofumi replied:
To get /apis/rest/items, we need this for client context. So I think it is still useful RestUrl for easier and reliable way

```python
        with current_app.test_request_context():
            return url_for('bepasty_apis.items')
```

ThomasWaldmann replied:
if the tests needing a request context would "require" it via a fixture param, guess this RestUrl class could be removed and we could directly use url_for() inside the test function.

OGAWAHirofumi replied:
this test_request_context() is **fake** request context, just to make url_for() works. so I don't want put other code under this context.

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+    assert len(response.json) == 1
+
+    fid = list(response.json.keys())[0]
+    # meta = list(response.json.values())[0]
```
?

OGAWAHirofumi replied:
It was reminder to get meta. removed

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+    filename = 'test.py'
+    ftype = 'text/x-python'
+
+    # upload a first half of data
```
upload first part of data

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+                       ftype=ftype, range_str=range_str)
+    check_upload_response(response, 200)
+
+    # upload a rest of data
```
upload remainder of data

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented on src/bepasty/apis/rest.py:
```diff
@@ -15,7 +16,44 @@
 
 
-class ItemUploadView(MethodView):
+# This wrapper to handle exception of REST api.
+# @blueprint.add_errorhandler decorator can do it though. We have
+# "/lodgeit/" in same blueprint, and didn't find the way to exclude
+# "/lodgeit/" from add_errorhandler to handle exception as web site,
+# not REST api.
```
Is this what you mean?
```
This wrappper handles exceptions in the REST api implementation.

The @blueprint.add_errorhandler decorator could do this, but we have
"/lodgeit/" in the same blueprint and there is no way to exclude this from
using the same error handler.
We can't use the same error handler though, because ...
```

OGAWAHirofumi replied:
thanks. fixed as following

```python
# This wrappper handles exceptions in the REST api implementation.
#
# The @blueprint.add_errorhandler decorator could do this, but we have
# "/lodgeit/" in the same blueprint and there is no way to exclude
# from using the same error handler ("/lodgeit/" is not REST api).
```

ThomasWaldmann commented on src/bepasty/apis/rest.py:
```diff
@@ -171,77 +172,52 @@ def get(self):
         else:
             if request_range.end == -1:
-                range_end = item.data.size - 1
+                size = item.data.size
```
guess this needs to consider request_range.begin

ThomasWaldmann replied:
in general, if the old way of dealing with begin/end was **not** broken, i'ld prefer you to not change it, but just keep it as is.

OGAWAHirofumi replied:
yes, usually I dont change. but this time self.stream() was consolidated and was changed to handle web site.  so -1 and +1 was repeated, and unreadable for me.

anyway, reverted unnecessary changes.

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -477,6 +477,22 @@ def test_upload_range(client_fixture):
+                       ftype=ftype, range_str=range_str, no_encode=True)
+    check_err_response(response, 400)
+
+    # server should not left garbage files
```
... must not have left ...

OGAWAHirofumi replied:
thanks. fixed

ThomasWaldmann commented on src/bepasty/apis/rest.py:
```diff
@@ -54,6 +54,52 @@ def post(self, *args, **kwargs):
+            raw_data = base64_b64decode(request.data)
+            file_data = BytesIO(raw_data)
+        except (base64.binascii.Error, TypeError):
+            return 'Data body could not decode', 400
```
'Could not decode data body'

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -224,7 +224,7 @@ def test_config(client_fixture):
 
 def _upload(client, data, token=None, filename=None, ftype=None, lifetime=None,
-            range_str=None, trans_id=None, no_encode=False):
+            range_str=None, trans_id=None, no_encode=False, no_range=False):
```
again: avoid double negation. `range=True`

OGAWAHirofumi replied:
fixed, changed to set_range

ThomasWaldmann replied:
ah, right, better not overwrite useful builtins.

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -240,7 +240,8 @@ def _upload(client, data, token=None, filename=None, ftype=None, lifetime=None,
     if range_str is None:
         range_str = 'bytes 0-{}/{}'.format(payload_len - 1, payload_len)
-    headers['Content-Range'] = range_str
+    if not no_range:
```
if range: ...

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented on src/bepasty/utils/date_funcs.py:
```diff
@@ -10,7 +11,10 @@ def get_maxlife(data, underscore):
+    try:
+        value = int(data.get(value_key, value_default))
+    except (ValueError, TypeError):
+        raise BadRequest(description='Maxlife-Value header is incorrect')
```
guess this should use value_key to generate the description.

OGAWAHirofumi replied:
fixed as following

```python
        field = 'maxlife_value' if underscore else 'Maxlife-Value'
        raise BadRequest(description='{} header is incorrect'.format(field))
```

ThomasWaldmann commented on src/bepasty/utils/date_funcs.py:
```diff
@@ -31,6 +35,8 @@ def time_unit_to_sec(value, unit):
         'FOREVER': FOREVER,
     }
+    if units.get(unit) is None:
+        raise BadRequest(description='Maxlife-Unit header is incorrect')
```
this function does not really know what it is used for, so it should rather raise a KeyError and let the caller generate the BadRequest.

OGAWAHirofumi replied:
hm, time_unit_to_sec() is only used from get_maxlife() if exclude tests/. it becomes like following. this was your intent?

```python
try:
    get_maxlife(...)
except KeyError:
     raise BadRequest(description='Maxlife-Unit header is incorrect')
```

ThomasWaldmann replied:
yes, or use some variable, if that can have underscore alternatively.

OGAWAHirofumi replied:
ah, ok. changed the get_maxlife() to catch KeyError, and time_unit_to_sec() keep as is (KeyError is raised naturally if no key)

```python
    try:
        return time_unit_to_sec(value, unit)
    except KeyError:
        field = unit_key if underscore else 'Maxlife-Unit'
        raise BadRequest(description='{} header is incorrect'.format(field))
```

ThomasWaldmann replied:
ok

ThomasWaldmann commented on src/bepasty/utils/upload.py:
```diff
@@ -36,9 +37,12 @@ def filter_size(cls, i):
+        try:
+            i = int(i)
+        except (ValueError, TypeError):
+            raise BadRequest(description='Size is incorrect')
```
Size is invalid

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented on src/bepasty/apis/rest.py:
```diff
@@ -157,9 +158,17 @@ def post(self):
+            try:
+                name_b = base64_b64decode(trans_id_b)
+            except (base64.binascii.Error, TypeError):
+                raise BadRequest(description='{} is incorrect'.format(TRANSACTION_ID))
```
do you mean: `could not decode {}` ?

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented on src/bepasty/apis/rest.py:
```diff
@@ -157,9 +158,17 @@ def post(self):
+                item = current_app.storage.openwrite(name)
+            except (OSError, IOError) as e:
+                if e.errno == errno.ENOENT:
+                    raise BadRequest(description='{} is incorrect'.format(TRANSACTION_ID))
```
guess this should be something like "could not find storage item for transaction id"

OGAWAHirofumi replied:
fixed

ThomasWaldmann commented:
Note: this PR was still too large overall.

I appreciate you putting lots of work into improving bepasty, but please do smaller and more focussed PRs.

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+        self.now = now
+        self.set_time(self.now)
+
+    def set_time(self, now=None):
+        self.now = now
+        if now and time.time != self.get_time:
+            time.time = self.get_time
+        elif now is None and time.time == self.get_time:
+            time.time = self.orig
```
faketime.set_time() is used like this too.

```python
        faketime.set_time(3600 * 24 * 365 * 2)
        check_expired(client, fid, download=True)
        # check again (should be deleted already)
        faketime.set_time(100)
```

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+
+
+def wait_background():
+    """Wait background threads"""
```
fixed

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+
+
+class TmpConfig:
+    def __init__(self, app, tmp_config={}):
```
fixed

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+    def __init__(self, now=None):
+        self.orig = time.time
+        self.now = now
+        self.set_time(self.now)
```
this is also used as following too, or I'm missing something

```python
        with FakeTime(int(time.time()) + i):
            meta = upload(client, data, token='full', filename=filename,
                          ftype=ftype, lifetime=lifetime)
```

ThomasWaldmann replied:
well, if you use a CM, the enter method is called when entering the block inside the CM.

so, faking time can get initialised when entering and normality restored when exiting.

OGAWAHirofumi replied:
It was for

```python
faketime = FakeTIme(100)
```

usage. but this usage was gone from current code. so removed `.set_time()` call from `__init__`.

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+
+
+def add_auth(user, password, headers=None):
+    if headers is None:
+        headers = dict()
```
fixed

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+def add_auth(user, password, headers=None):
+    if headers is None:
+        headers = dict()
+    headers.update({'Authorization': _basic_auth_str(user, password)})
```
fixed

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+def test_invalid_url(client_fixture):
+    _, client, _ = client_fixture
+
+    # invalid url
+    response = client.get('/apis/rest/foo')
```
fixed

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+                        check_data=True):
+    ftype = meta['file-meta'][TYPE]
+    filename = meta['file-meta'][FILENAME]
+    if offset is None:
+        offset = 0
```
fixed

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+    if total_size is None:
+        total_size = len(data)
+
+    dispoition = 'attachment; filename="{}"'.format(filename)
```
fixed

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+        assert data == response.data
+
+
+def check_json_response(response, metas, check_data=True):
```
this can be multiple of meta on list api

ThomasWaldmann replied:
ah

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+    check_response(response, code, 'text/html; charset=utf-8', check_data)
+
+    if code == 200:
+        assert len(response.headers[TRANSACTION_ID]) > 0
```
this meat, there is Transction-ID, and its length > 0

ThomasWaldmann replied:
ok

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+    check_response(response, 200)
+
+    # token auth (unknown user)
+    response = client.get(url.list + '?token=invalid')
```
fixed. looks like working with

```python
    with app.app_context():
        with current_app.test_request_context():
            return url_for('bepasty_apis.items_download', name=self.fid)
```

ThomasWaldmann replied:
great!

OGAWAHirofumi commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,666 @@
+
+
+def _upload(client, data, token=None, filename=None, ftype=None, lifetime=None,
+            range_str=None, trans_id=None, no_encode=False):
```
fixed

OGAWAHirofumi commented on src/bepasty/views/download.py:
```diff
@@ -13,58 +13,62 @@
+
+            if delete_if_lifetime_over(item, name):
+                raise NotFound()
+            need_close = False
```
yes. Since self.response() uses item.data, item should not be closed.

ThomasWaldmann replied:
ah, ok.

OGAWAHirofumi commented:
Hm, then all those patches are depending on previous patches. So if separated more than this, it has to wait the result of previous PR?

OGAWAHirofumi commented:
this is to show updated patches. If "separated PR and wait result" and "then fix or next PR" is prefered, I will try with new PR (or if there are better way to do it, it would be fine).

OGAWAHirofumi commented:
changes:

- caa507a: removed `self.set_time()` from `FakeTime().__init__()`
- b1b46ef: changed to check `KeyError` in `get_maxlife()`, and handled `underscore`

OGAWAHirofumi commented:
changes:
    
- ae36009: use copy.deepcopy() in TmpConfig

(included in previous update)
- caa507a: removed self.set_time() from FakeTime().__init__()
- b1b46ef: changed to check KeyError in get_maxlife(), and handled underscore

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,817 @@
+        wait_background()
+
+
+class TmpConfig:
+    def __init__(self, app, tmp_config=None):
+        self.config = app.config
+        self.orig_config = copy.deepcopy(app.config)
+        self.tmp_config = {} if tmp_config is None else tmp_config
+
+    def __enter__(self):
+        self.config.update(self.tmp_config)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.config.update(self.orig_config)
```
do we need self.config? guess directly using app.config is less confusing:

OGAWAHirofumi replied:
hm, then we have to make app global? IMO, self.config looks better.

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,817 @@
+
+
+class RestUrl:
+    def __init__(self, fid=None):
+        self.fid = fid
```
minor nitpick: `fid` is not a very speaking name. bepasty usually talks about items, so how about `item_id`?

OGAWAHirofumi replied:
i wanted to use shorter name though. changed to item_id

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,817 @@
+
+        with app.test_client() as client:
+            with app.app_context():
+                yield app, client, faketime
```
hmm, would it be easier to have separate fixtures for these?

- app
- client
- request
- faketime

OGAWAHirofumi replied:
easier would be one. because no need to think/handle other combinations.

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,817 @@
+        assert int(response.headers['Content-Length']) == len(response.data)
+
+
+re_html_tag = re.compile(r'<.+>')
```
if this is only used in the function below, move it into that scope. we don't need to micro-optimize here.

OGAWAHirofumi replied:
changed

ThomasWaldmann commented on src/bepasty/utils/date_funcs.py:
```diff
@@ -1,17 +1,26 @@
+    try:
+        value = int(data.get(value_key, value_default))
+    except (ValueError, TypeError):
+        field = value_key if underscore else 'Maxlife-Value'
+        raise BadRequest(description='{} header is incorrect'.format(field))
```
remove 17 and just use value_key instead of field?

OGAWAHirofumi replied:
other header fields uses camel case. if you don't care inconsistency, value_key would be fine.

ThomasWaldmann replied:
i think 17 is just inconsistent.

we access via value_key, always.

so the error message should not use value_key for one case and something hardcoded for the other.

OGAWAHirofumi replied:
users of get_maxlife() are, one is request.form and one is request.headers.

And questionable one is request.headers, because it is case insensitive dict, and the header is usually using camel case. If we don't care this camel case in request.headers,

for example,

```python
        if not request.headers.get("Content-Range"):
            return 'Content-Range not specified', 400
```

Content-Range error uses camel case. So actually I myself don't care lower or camel though, I wanted to confirm you also don't care this.

ThomasWaldmann replied:
don't care, we'll just use the key (which we really use to access the dict) also for the message.

OGAWAHirofumi replied:
the idea to remove "field = ..." stuff, how about the following change? using camel case for *_key

```python
def get_maxlife(data, underscore):
    unit_key = 'maxlife_unit' if underscore else 'Maxlife-Unit'
    unit_default = 'MONTHS'
    unit = data.get(unit_key, unit_default).upper()
    value_key = 'maxlife_value' if underscore else 'Maxlife-Value'
    value_default = '1'
    try:
        value = int(data.get(value_key, value_default))
    except (ValueError, TypeError):
        raise BadRequest(description='{} header is incorrect'.format(value_key))
    try:
        return time_unit_to_sec(value, unit)
    except KeyError:
        raise BadRequest(description='{} header is incorrect'.format(unit_key))
```

ThomasWaldmann replied:
i guess within the current commits in this PR, you have the least possible changes, so keep it as is.

in general, i do not like huge pull requests and i especially do not like even more huge pull requests that additionally "fix" (change) stuff that was **not** broken.

the work you do within this PR is great and very welcome (esp. that you added a lot of tests), but let's concentrate on real fixes here.

OGAWAHirofumi replied:
I agree, but at same time disagree. making clean history is good and very preferred, however, it takes more time and cost. so we would need a balance as usual. and IMO, at least, making forward and better result should be more important than cleaner history.

anyway, i already reverted those changes.

ThomasWaldmann commented on src/bepasty/utils/date_funcs.py:
```diff
@@ -1,17 +1,26 @@
+    try:
+        return time_unit_to_sec(value, unit)
+    except KeyError:
+        field = unit_key if underscore else 'Maxlife-Unit'
+        raise BadRequest(description='{} header is incorrect'.format(field))
```
remove 22 and just use unit_key instead of field?

OGAWAHirofumi replied:
same with value_key

ThomasWaldmann replied:
same here.

OGAWAHirofumi commented:
changes:
- 61c3e3d: rename fid to item_id
- c2b5f34: use re.match simply

(included in previous update)
- ae36009: use `copy.deepcopy()` in `TmpConfig`
- caa507a: removed `self.set_time()` from `FakeTime().__init__()`
- b1b46ef: changed to check `KeyError` in `get_maxlife()`, and handled `underscore`

OGAWAHirofumi commented:
ping?

ThomasWaldmann commented on src/bepasty/utils/_compat.py:
```diff
@@ -14,9 +15,15 @@
     range_type = xrange  # noqa: F821
     bytes_type = str
     iteritems = lambda d: d.iteritems()  # noqa: E731
+
+    def base64_b64decode(data):
+        return base64.b64decode(data)
 else:
     from urllib.parse import urlparse, urljoin
 
     range_type = range
     bytes_type = bytes
     iteritems = lambda d: iter(d.items())  # noqa: E731
+
+    def base64_b64decode(data):
+        return base64.b64decode(data, validate=True)
```
Maybe this is well-meant (like "validating is always better than not"), but in the end, this implements different behaviour for python 2.7 and python 3.x. I don't think we want different behaviour.

So, there should be a ticket about changing this to validating **after** we dropy python 2 support, but this should not be changed within this PR.

OGAWAHirofumi replied:
Hm, maybe no. in my understand, python 2.7 doesn't have validate parameter and validate by default. but python3 default is no validation.

So this sets validate to make be same behavior.

ThomasWaldmann replied:
no, 2.7 does not validate by default, it strips invalid stuff, like default in 3.x - see the docs.

OGAWAHirofumi replied:
- https://docs.python.org/2.7/library/base64.html
The decoded string is returned. A TypeError is raised if s is incorrectly padded. Characters that are neither in the normal base-64 alphabet nor the alternative alphabet are discarded prior to the padding check.

- https://docs.python.org/3/library/base64.html
If validate is False (the default), characters that are neither in the normal base-64 alphabet nor the alternative alphabet are discarded prior to the padding check. If validate is True, these non-alphabet characters in the input result in a binascii.Error.

Looks like it rasise TypeError?

ThomasWaldmann replied:
characters being discarded is the common default, so both behave the same.

OGAWAHirofumi replied:
hm, sure. looks like I somehow confused when testing.

OGAWAHirofumi commented:
changes:
- removed compat base64_b64decode completely

(included in previous update)
- 61c3e3d: rename fid to item_id
- c2b5f34: use re.match simply
- ae36009: use `copy.deepcopy()` in `TmpConfig`
- caa507a: removed `self.set_time()` from `FakeTime().__init__()`
- b1b46ef: changed to check `KeyError` in `get_maxlife()`, and handled `underscore`

ThomasWaldmann commented on src/bepasty/apis/rest.py:
```diff
@@ -5,17 +5,103 @@
+# "/lodgeit/" in the same blueprint and there is no way to exclude
+# from using the same error handler ("/lodgeit/" is not REST api).
+def rest_errorhandler(func):
+    def error_message(description, code):
+        # maybe better to return error as json?
+        return description, code
```
remove this function

ThomasWaldmann commented on src/bepasty/apis/rest.py:
```diff
@@ -5,17 +5,103 @@
+        try:
+            return func(*args, **kwargs)
+        except HTTPException as e:
+            return error_message(e.description, e.code)
```
```
... as exc:
    pass
```

ThomasWaldmann commented on src/bepasty/apis/rest.py:
```diff
@@ -5,17 +5,103 @@
+                # if testing/debug mode, re-raise
+                raise
+            exc = InternalServerError()
+            return error_message(exc.description, exc.code)
```
outdent this line 1 level and use:
```
return exc.description, exc.code
```

OGAWAHirofumi replied:
you meant the following?

```python
def rest_errorhandler(func):
    def handler(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except HTTPException as exc:
            pass
        except Exception:
            if current_app.propagate_exceptions:
                # if testing/debug mode, re-raise
                raise
            exc = InternalServerError()
        return exc.description, exc.code

    return handler
```

if so, looks like not working,
```
src/bepasty/apis/rest.py:28:9: F841 local variable 'exc' is assigned to but never used
```

ThomasWaldmann replied:
ehrm. either i do not know some strange scoping rule or this is a false positive from flake8 tester...

OGAWAHirofumi replied:
it was actual error in python3 at least

ThomasWaldmann replied:
ok, looks like the first. exc is only defined inside the except block.

```
def rest_errorhandler(func):
    def handler(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except HTTPException as exc:
            return exc.description, exc.code
        except Exception:
            if current_app.propagate_exceptions:
                # if testing/debug mode, re-raise
                raise
            exc = InternalServerError()
            return exc.description, exc.code

    return handler
```

OGAWAHirofumi commented:
changes:
- removed error_message() from rest_errorhandler
- removed "field = ..." and use camel case for get_maxlife()

(included in previous update)
- removed compat base64_b64decode completely
- 61c3e3d: rename fid to item_id
- c2b5f34: use re.match simply
- ae36009: use `copy.deepcopy()` in `TmpConfig`
- caa507a: removed `self.set_time()` from `FakeTime().__init__()`
- b1b46ef: changed to check `KeyError` in `get_maxlife()`, and handled `underscore`

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,816 @@
+import copy
+import hashlib
+import re
+from requests.auth import _basic_auth_str
```
this does not look like a public function, can you search for a better way?

OGAWAHirofumi replied:
I cant guess your favorite though, how about this?

```python
def add_auth(user, password, headers=None):
    req = Request('POST', 'http://dummy', auth=(user, password))

    headers = headers if headers is not None else {}
    headers['Authorization'] = req.prepare().headers['Authorization']
    return headers
```

ThomasWaldmann replied:
constructing a full Request ovjects seems a bit much, I thought there should be some public function somewhere that does what we need.

but functions starting with single or double underscore are usually not meant to be a public stable API, so that usually means "don't use this from outside".

ThomasWaldmann replied:
i looked around and found it rather embarrassing to not being able to find a reusable implementation of that in python stdlib, werkzeug, requests. everywhere it is just hardcoded inside some other stuff.

so, ignore my feedback and just use that internal function until we find something better.

OGAWAHirofumi replied:
ok, i will make PR with _basic_auth_str later.

ThomasWaldmann replied:
thanks!

ThomasWaldmann commented on src/bepasty/utils/date_funcs.py:
```diff
@@ -1,17 +1,24 @@
 def get_maxlife(data, underscore):
-    unit_key = 'maxlife_unit' if underscore else 'maxlife-unit'
-    unit_default = 'MONTH'
+    unit_key = 'maxlife_unit' if underscore else 'Maxlife-Unit'
+    unit_default = 'MONTHS'
     unit = data.get(unit_key, unit_default).upper()
-    value_key = 'maxlife_value' if underscore else 'maxlife-value'
+    value_key = 'maxlife_value' if underscore else 'Maxlife-Value'
```
if maxlife-unit and maxlife-value were not broken, do not "fix" them.

OGAWAHirofumi replied:
ok, if you think so. removed

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,816 @@
+                       ftype=ftype, range_str=range_str)
+    check_upload_response(response, 200)
+
+    # upload reminder of data
```
remainder (please fix globally)

OGAWAHirofumi replied:
fixed. (there was no others)

ThomasWaldmann commented on src/bepasty/tests/test_rest_server.py:
```diff
@@ -0,0 +1,816 @@
+
+    # upload reminder of data
+    data = UPLOAD_DATA[sep:]
+    range_str = 'bytes {}-{}/{}'.format(sep, len(UPLOAD_DATA) - 1,
```
shouldn't this be: `sep, sep + len(UPLOAD_DATA) - 1, len(UPLOAD_DATA)`

and if so, why did the test succeed? bug?

ThomasWaldmann replied:
ah, no, my fault. it would be sep + len(data) - 1, but your version is also correct.

OGAWAHirofumi commented:
changes:
- removed Maxlife-*
- fix typo reminder => remainder
- try other way for basic auth

(included in previous update)
- removed error_message() from rest_errorhandler
- removed "field = ..." and use camel case for get_maxlife()
- removed compat base64_b64decode completely
- 61c3e3d: rename fid to item_id
- c2b5f34: use re.match simply
- ae36009: use `copy.deepcopy()` in `TmpConfig`
- caa507a: removed `self.set_time()` from `FakeTime().__init__()`
- b1b46ef: changed to check `KeyError` in `get_maxlife()`, and handled `underscore`

ThomasWaldmann commented:
Note: added this to 0.6.0 (next release) milestone.

OGAWAHirofumi commented:
changes:
- reverted "try other way for basic auth", now use _basic_auth_str()

(included in previous update)
- removed Maxlife-*
- fix typo reminder => remainder
- removed error_message() from rest_errorhandler
- removed "field = ..." and use camel case for get_maxlife()
- removed compat base64_b64decode completely
- 61c3e3d: rename fid to item_id
- c2b5f34: use re.match simply
- ae36009: use `copy.deepcopy()` in `TmpConfig`
- caa507a: removed `self.set_time()` from `FakeTime().__init__()`
- b1b46ef: changed to check `KeyError` in `get_maxlife()`, and handled `underscore`

ThomasWaldmann commented:
Yay, merged! Thanks for your PR!
